app/models/locations/map.rb

```diff
@@ -0,0 +1,90 @@
+module Locations
+  class Map
+    include ActiveModel::Model
+
+    def initialize(params = {})
+      @@locator_url = "http://library.princeton.edu/locator/index.php?"
+      @@stackmap_url = "http://princeton.stackmap.com/view/?"
+      @params = params
+      @holding_location = set_holding_location(@params[:loc])
+      @bibrec = set_bibrec(@params[:id])
+      @lib = set_lib(@holding_location.locations_library_id)
+      @valid = is_valid
+    end
+
+    def url
+        if @valid
+          if !@holding_location.open
```

Comment:
Better not to use instance variables in methods - accessors are better for messages. Even just `attr_reader :holding_location` at the top should fix that.